<a href="https://colab.research.google.com/github/aayomide/hamoye-data-engineering/blob/main/stage_a/stage_a_quiz_answers.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### STAGE A - PYTHON FOR MACHNE LEARNING QUIZ ANSWERS

In [ ]:
# import libraries
import pandas as pd
import numpy as np

In [ ]:
file_url = 'https://github.com/HamoyeHQ/HDSC-Introduction-to-Python-for-machine-learning/files/7768140/FoodBalanceSheets_E_Africa_NOFLAG.csv'

df_food = pd.read_csv(file_url, encoding='latin-1')
df_food.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

In [ ]:
# check file info
print(df_food.shape)
df_food.info()

(60943, 12)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60943 entries, 0 to 60942
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Area Code     60943 non-null  int64  
 1   Area          60943 non-null  object 
 2   Item Code     60943 non-null  int64  
 3   Item          60943 non-null  object 
 4   Element Code  60943 non-null  int64  
 5   Element       60943 non-null  object 
 6   Unit          60943 non-null  object 
 7   Y2014         59354 non-null  float64
 8   Y2015         59395 non-null  float64
 9   Y2016         59408 non-null  float64
 10  Y2017         59437 non-null  float64
 11  Y2018         59507 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 5.6+ MB


In [ ]:
# Question 1
# What is the mean and standard deviation across the whole dataset for the year 2017 to 2 decimal places?

df_food['Y2017'].describe().round(2)

count     59437.00
mean        140.92
std        1671.86
min       -1582.00
25%           0.00
50%           0.10
75%           9.00
max      190873.00
Name: Y2017, dtype: float64

> The mean and the standard deviation across the whole dataset for the year 2017 is 140.92 and 1671.86 respectively.

In [ ]:
# Question 2
# What is the total number and percentage of missing data in 2014 to 3 decimal places?

(
    df_food['Y2014'].isnull().sum() ,
    round(df_food['Y2014'].isnull().mean() * 100, 3)
)

(1589, 2.607)

> The total number and percentage of missing data in 2014 is 1589 and 2.607% respectively.

In [ ]:
# Question 3
# Given the following python code, what would the output of the code give?

my_tuppy = (1,2,5,8)

my_tuppy[2] = 6

TypeError: 'tuple' object does not support item assignment

In [ ]:
# Question 4
# What is the total number of unique countries in the dataset?

df_food['Area'].nunique()

49

In [ ]:
# Question 5
# A pandas Dataframe with dimensions (100,3) has how many features and observations?

#answer: 3 features and 100 observations

In [ ]:
# Question 7
# Perform a groupby operation on ‘Element’.  What is the total number of the sum of Processing in 2017?

df_food.groupby('Element')[['Y2017']].sum()\
        .loc['Processing']

Y2017    292836.0
Name: Processing, dtype: float64

> The total sum of processing in 2017 is 292836.

In [ ]:
# Question 8
# Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the 7th lowest sum in 2017?

df_food.groupby('Area')['Y2017'].sum()\
    .sort_values()\
    .reset_index()\
    .iloc[6]

Area     Guinea-Bissau
Y2017         19102.77
Name: 6, dtype: object

> The Area with the 7th lowest sum in year 2017 is Guinea-Bissau.

In [ ]:
# Question 10
# What would be the output for?

S = [['him', 'sell'], [90, 28, 43]]

S[0][1][1]

'e'

In [ ]:
# Question 11
# Perform a groupby operation on ‘Element’.  What year has the highest sum of Stock Variation?

df_food.groupby('Element')[['Y2014', 'Y2015', 'Y2016', 'Y2017', 'Y2018']].sum()\
        .loc['Stock Variation']\
        .sort_values(ascending=False)

Y2014    58749.83
Y2017    54316.91
Y2015    34910.99
Y2016    33140.12
Y2018    20577.91
Name: Stock Variation, dtype: float64

> The result shows that the Year 2014 had the highest sum of Stock Variation.

In [ ]:
# Question 13
# Answer the following questions based on the African food production dataset provided by the FAO website already provided
# What is the total sum of Wine produced in 2015 and 2018 respectively?
# Hint: Perform a groupby sum aggregation on ‘Item’

df_food.groupby('Item')[['Y2015', 'Y2018']].sum()\
        .loc['Wine']

Y2015    4251.81
Y2018    4039.32
Name: Wine, dtype: float64

In [ ]:
# Question 14
# Which year had the least correlation with ‘Element Code’?

df_food.select_dtypes(include='number').corr()\
        ['Element Code']\
        .sort_values()

Item Code      -0.024683
Area Code      -0.000209
Y2016           0.023444
Y2015           0.023889
Y2017           0.024254
Y2018           0.024279
Y2014           0.024457
Element Code    1.000000
Name: Element Code, dtype: float64

> The year with the least correlation with Element Code is 2016.

In [ ]:
# Question 15
# Select columns ‘Y2017’ and ‘Area’, Perform a groupby operation on ‘Area’.  Which of these Areas had the highest sum in 2017?

df_food.groupby('Area')['Y2017'].sum()\
    .sort_values(ascending=False)\
    .head(1)

Area
Nigeria    1483268.23
Name: Y2017, dtype: float64

> The Area with the highest sum in 2017 was Nigeria.

In [ ]:
# Question 16
# Given the following numpy array

array  = np.array(
    [ [94, 89, 63],
      [93, 92, 48],
      [92, 94, 56] ]
)

# How would you select  the elements in bold and italics from the array?

array[:2,1:]

array([[89, 63],
       [92, 48]])

In [ ]:
# Question 17
# If you have the following list

lst = [[35, 'Portugal', 94], [33, 'Argentina', 93], [30 , 'Brazil', 92]]

col = ['Age', 'Nationality', 'Overall']

# How do you create a pandas DataFrame using this list, to look like the table below?

In [ ]:
pd.DataFrame(lst, columns=col, index=[i for i in range(1,4)])

Age Nationality  Overall
1   35    Portugal       94
2   33   Argentina       93
3   30      Brazil       92

In [ ]:
# Question 18
# Which of these python data structures is unorderly?

# answer: Set and Dictionary

In [ ]:
# Question 19
# What is the total Protein supply quantity in Madagascar in 2015?

# df_food[(df_food.Element == 'Protein supply quantity (g/capita/day)') & (df_food.Area == "Madagascar")]
df_food[(df_food.Element.str.contains('Protein')) & (df_food.Area == 'Madagascar')]\
        ['Y2015']\
        .sum()\
        .round(2)

173.05

> The total protein supply quantity in Madagascar in 2015 is 173.05 g/capita/day.

In [ ]:
# Question 20
# Consider the following list of tuples:

y = [(2, 4), (7, 8), (1, 5, 9)]

# How would you assign element 8 from the list to a variable x?
y[1][1]

8